### PR TITLE
i#1578 arm detach: Restore TLS register on detach

### PR DESCRIPTION
**Pull request by derekbruening**
Sets os_should_swap_state() to true.
Adds support for os_swap_context() twice in a row with the same state.
Removes os_swap_context_go_native().

The api.detach test now passes on AArch64.

Enables all of the api.* and api.static_* tests for AArchXX, except
api.startstop( crash) and api.static_prepop (hang) which will be fixed
separately.

Fixes a build error in the api.static_crash test: a missing header.

Issue: #1578, #1582

### CODE REVIEW

**Comment by derekbruening** [comment]
Will update the description to include not enabling api.static_noclient and api.thread_churn.

**Comment by derekbruening** [comment]
The A64 failure is odd: it lists `code_api|common.allasm_aarch64_isa` as a failure but it passed in the output above.

File: suite/tests/CMakeLists.txt
Comment (abhinav92003) [inline suite/tests/CMakeLists.txt:2779]: Is the crash mentioned here related to the `XXX i#1578` comment in `core/unix/os.c`?

**Review by abhinav92003 — APPROVED** [review]

File: suite/tests/CMakeLists.txt
Comment (abhinav92003) [reply suite/tests/CMakeLists.txt:2779]: Oh that comment is just for pure-asm apps. So probably not.

Comment (derekbruening) [reply suite/tests/CMakeLists.txt:2779]: The os.c comment about no-libc apps?  No, that was affecting the allasm tests which have no TLS of their own b/c they don't use any libraries.

**Comment by derekbruening** [comment]
> The A64 failure is odd: it lists `code_api|common.allasm_aarch64_isa` as a failure but it passed in the output above.

I figured it out: the release build test passed, and the web interface truncated the start of the log.  On expanding it we see the failure in debug.  It's the same as what I fixed with the os.c no-libc check for having a TLS segment set up, but this test puts a garbage value in the thread register (just as part of testing the ISA) so we have to rule that out too.

File: core/unix/os.c
Comment (petrochenkov) [inline core/unix/os.c:2708]: @derekbruening 
I found this change to be the source of a regression on Android - an application run under DR (without any extra options or clients) is now failing with "Cannot correctly handle received signal 11 in thread X.".

`INTERNAL_OPTION(private_loader)` here reports `true` on Android, but private loader doesn't really work there (cc https://github.com/DynamoRIO/dynamorio/issues/3543), and in this particular case something breaks if it's considered working.

I guess, since things generally work on Android, considering `INTERNAL_OPTION(private_loader)` true on Android is not *too* harmful, but apparently there's an unknown set of cases where it is indeed harmful.
Not sure what is the best course of action here, besides actually implementing the private loader.

Comment (derekbruening) [reply core/unix/os.c:2708]: Hmm, the real problem is that there is no CI testing for Android: #4639.  The code reflects the state as of several years ago, when the private loader did work.  Like you said, the best solution would be to get it working with the latest Android, though if you have a quick fix that improves things in the short term please submit a PR.

Unfortunately while the handful of regular DR maintainers are happy to provide advice and consulting for Android we are not likely to be able to spend time on actual coding work targeting Android.  Several people have expressed interest in Android support; hopefully someone will contribute some time.

Comment (derekbruening) [reply core/unix/os.c:2708]: @petrochenkov if you could file a new issue on the regression; and ideally submit a PR with a quick fix if it's blocking things.

Comment (petrochenkov) [reply core/unix/os.c:2708]: The quick fix that seems to work in practice is to continue returning `false` in the `defined(ANDROID)` case, unfortunately I cannot submit it without an approval.

Comment (derekbruening) [reply core/unix/os.c:2708]: > The quick fix that seems to work in practice is to continue returning false in the defined(ANDROID) case, unfortunately I cannot submit it without an approval.

You mean some approval process from your employer I assume.

**Pull request merged**
